docs: update spec for Row Variables and no-input-extensions (#1402)
CQCL/hugr: specification/hugr.md

classes are distinguished by whether the runtime values of those types can be implicitly
copied or discarded (multiple or 0 links from on output port respectively):

> - For the broadest class (`AnyType`), the only operation supported is the identity operation (aka no-op, or `lift` - see [Extension Tracking](#extension-tracking) below). Specifically, we do not require it to be possible to copy or discard all values, hence the requirement that outports of linear type must have exactly one edge. (That is, a type not known to be in the copyable subset).

Author: This was a driveby commoning up, see below

Type ::= Sum([#]) -- disjoint union of rows of other types, tagged by unsigned int
       | Opaque(Name, [TypeArg]) -- a (instantiation of a) custom type defined by an extension
       | Function(#, #, Extensions) -- monomorphic function: arguments, results, and delta (see below)

>        | Variable -- refers to a TypeParam bound by an the nearest enclosing FuncDefn node or polymorphic type scheme

Author: This dates back to #906!

Author: Oops, `by an the` back to `by an`

  For example, `Type::Function(usize, unit, <exts>)` is equivalent shorthand
  for `Type::Function(#(usize), #(unit), <exts>)`.
* When a `TypeArg::Sequence` is provided as argument for such a TypeParam, we allow
  elements to be a mixture of both types (including variables of kind
  `TypeParam::Type(_)`) and also row variables. When such variables are instantiated
  (with other Sequences) the elements of the inner Sequence are spliced directly into
  the outer (concatenating their elements), eliding the inner (Sequence) wrapper.

For example, a polymorphic FuncDefn might declare a row variable X of kind
`TypeParam::List(TypeParam::Type(Copyable))` and have as output a (tuple) type
`Sum([#(X, usize)])`. A call that instantiates said type-parameter with
`TypeArg::Sequence([usize, unit])` would then have output `Sum([#(usize, unit, usize)])`.

See [Declarative Format](#declarative-format) for more examples.

Note that since a row variable does not have kind Type, it cannot be used as the type of an edge.

### Extension Tracking

The type of `Function` includes a set of [extensions](#extension-system) which are required to execute the graph.

> Similarly, every dataflow node in the HUGR has a set of extensions required to execute the node (computed from its operation), also known as the "delta". The delta of any node must be a subset of its parent,

Reviewer: ```suggestion
Similarly, every dataflow node in the HUGR has a set of extensions required to execute the node (computed from its operation), also known as the "delta". The delta of any node must be a subset of its parent's delta,
```

except for FuncDefn's:

> * the delta of any child of a FuncDefn, must be a subset of the extensions in the FuncDefn's *type*

Author: This is a bit awkward. Suggestions welcome

Reviewer: just remove the comma I think

about where to run different parts of the graph by inspecting their
extension requirements.

Special operations **lift** and **liftGraph** can add extension requirements:
* `lift<E: ExtensionId, R: List<Type>>` is a node with input and output rows `R` and extension-delta `{E}`
* `liftGraph<N: ExtensionSet, I: List<Type>, E: ExtensionSet, O: List<Type>>` has one input
$ \vec{I}^{\underrightarrow{\;E\;}}\vec{O} $ and one output $ \vec{I}^{\underrightarrow{\;E \cup N\;}}\vec{O}$.
That is, given a graph, it adds extensions $N$ to the requirements of the graph.

The latter is useful for higher-order operations such as conditionally selecting
one function or another, where the output must have a consistent type (including
the extension-requirements of the function).

> # Having these as explicit nodes on the graph allows us to search for the

Author: Any point in keeping these sections I've commented out?

Reviewer: don't think so

    # outputs would be, in principle: Array<i+j>(t)
    # - but default type scheme interpreter does not support such addition
    # Hence, no signature block => will look up a compute_signature in registry.
  - name: TupleConcat
    description: "Concatenate two tuples"
    params:
      a: List[Type]
      b: List[Type]
    signature:
      inputs: Sum([a]), Sum([b])

>       outputs: Sum([(a,b)]) # Syntax! Sum([a,b]) would be either-a-or-b, this has one variant

Author: Bleurgh. It'd be awkward to write these examples *not* in the "declarative format", but OTOH I don't think we are really moving in this direction, and it's very unclear how to write these ATM.

Reviewer: I think better just to say what it is in the comment
```suggestion
      outputs: Sum([(a,b)]) # sum with a single variant of type (a, b)
```
